cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlManager.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol;
+
+import com.google.common.base.Strings;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.proto.sourcecontrol.AuthType;
+import io.cdap.cdap.proto.sourcecontrol.Provider;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryValidationFailure;
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.LsRemoteCommand;
+import org.eclipse.jgit.api.TransportCommand;
+import org.eclipse.jgit.api.TransportConfigCallback;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.api.errors.TransportException;
+import org.eclipse.jgit.internal.storage.file.FileRepository;
+import org.eclipse.jgit.lib.Ref;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+/**
+ * A Source Control Manager that is responsible for handling interfacing with git. It fetches credentials
+ * from a {@link io.cdap.cdap.api.security.store.SecureStore}, builds an {@link AuthStrategy} for different
+ * Git hosting services and provides version control operations.
+ */
+public class SourceControlManager implements AutoCloseable {
+  private static final Logger LOG = LoggerFactory.getLogger(SourceControlManager.class);
+  private static final String validationRepositoryDirectoryName = "validation";
+  private static final String localRepositoryDirectoryName = "local";
+  private final SourceControlContext context;
+  private final AuthStrategy authStrategy;
+  private final File localRepoDirectory;
+  private final File validationRepoDirectory;
+  private final int gitCommandTimeoutSeconds;
+  private Git git;
+
+  public SourceControlManager(SourceControlContext context) throws AuthStrategyNotFoundException {
```

Comment:
Removed the context class.

Reply:
I'm not sure what the Constructor can do when it doesn't know how to authenticate with the remote repository. It could set the auth strategy as null, but that would then cause an NPE if the request was incorrectly validated by the handler or if it came from somewhere else.

Reply:
I converted the exception into a RuntimeException in the constructor since it isn't expected to occur in the normal code path.